src/internal/kopia/upload_test.go

```diff
@@ -674,8 +755,867 @@ func (suite *HierarchyBuilderUnitSuite) TestBuildDirectoryTree_Fails() {
+				inbox.PrevPath = inboxPath
+				inbox.ColState = data.MovedState
+
+				work := mockconnector.NewMockExchangeCollection(newWorkPath, 0)
+				work.PrevPath = workPath
+				work.ColState = data.MovedState
+
+				return []data.Collection{inbox, work}
+			},
+			expected: &expectedNode{
+				name: testTenant,
+				children: []*expectedNode{
+					{
+						name: service,
+						children: []*expectedNode{
+							{
+								name: testUser,
+								children: []*expectedNode{
+									{
+										name: category,
+										children: []*expectedNode{
+											{
+												name: testInboxDir + "2",
+												children: []*expectedNode{
+													{
+														name: personalDir,
+														children: []*expectedNode{
+															{
+																name:     personalFileName1,
+																children: []*expectedNode{},
+															},
+															{
+																name:     personalFileName2,
+																children: []*expectedNode{},
+															},
+														},
+													},
+												},
+											},
+											{
+												name: workDir,
+												children: []*expectedNode{
+													{
+														name:     workFileName,
+														children: []*expectedNode{},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "MovesChildAfterAncestorDelete",
+			inputCollections: func(t *testing.T) []data.Collection {
+				newWorkPath := makePath(
+					t,
+					[]string{testTenant, service, testUser, category, workDir},
+				)
+
+				inbox := mockconnector.NewMockExchangeCollection(inboxPath, 0)
+				inbox.PrevPath = inboxPath
+				inbox.ColState = data.DeletedState
+
+				work := mockconnector.NewMockExchangeCollection(newWorkPath, 0)
+				work.PrevPath = workPath
+				work.ColState = data.MovedState
+
+				return []data.Collection{inbox, work}
+			},
+			expected: &expectedNode{
+				name: testTenant,
+				children: []*expectedNode{
+					{
+						name: service,
+						children: []*expectedNode{
+							{
+								name: testUser,
+								children: []*expectedNode{
+									{
+										name: category,
+										children: []*expectedNode{
+											{
+												name: workDir,
+												children: []*expectedNode{
+													{
+														name:     workFileName,
+														children: []*expectedNode{},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "ReplaceDeletedDirectory",
+			inputCollections: func(t *testing.T) []data.Collection {
+				personal := mockconnector.NewMockExchangeCollection(personalPath, 0)
+				personal.PrevPath = personalPath
+				personal.ColState = data.DeletedState
+
+				work := mockconnector.NewMockExchangeCollection(personalPath, 0)
+				work.PrevPath = workPath
+				work.ColState = data.MovedState
+
+				return []data.Collection{personal, work}
```

Comment:
Ordering is paramount here, yeah?  I assume that if the collection order were reversed: `{work, personal}`, then the work file would be added to personalDir first, then deleted along with the rest of the directory.  That poses some risks in terms of collection ordering; I don't recall (and maybe I'm just forgetting) that we've made any guarantees about the order of collections from backup data producers.  Do we need to sort collections by modTime for this to work as expected?

Reply:
nope, ordering does not matter at all since we first create the hierarchy and then send items to kopia. We may need to be a little more concerned about ordering if we ever start building the hierarchy in tandem with sending items to kopia, but we're not doing that right now